### PR TITLE
Attempt to unflake system metrics test for Windows

### PR DESCRIPTION
**Pull request by BenWilson2**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/BenWilson2/mlflow/pull/13773?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/13773/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 13773
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

The `test_automatic_system_metrics_monitor` suite has been flaky in Windows. Attempt slightly longer delay configurations for Windows due to the different VM runner config for Windows on GHA and the different behavior that Windows has with thread process management as compared to Linux.

### How is this PR tested?

- [X] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [X] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [X] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [X] No (this PR will be included in the next minor release)

### CODE REVIEW

**Review by TomeHirata — APPROVED** [review]
LGTM

**Comment by harupy** [comment]
@BenWilson2 can you run the test 1000 times make sure it's flaky?

File: tests/system_metrics/test_system_metrics_logging.py [hunk 1]
```diff
@@ -59,20 +60,23 @@ def test_manual_system_metrics_monitor():
     assert metrics_history[-1].step > 0
 
 
-def test_automatic_system_metrics_monitor():
+@pytest.mark.parametrize(
+    ("wait_time", "exit_wait_time", "polling_freq"), [(5, 3, 0.5) if is_windows() else (2, 2, 0.2)]
+)
```
Comment (harupy) [inline tests/system_metrics/test_system_metrics_logging.py:65]: We can set these parameter in the test.

File: tests/system_metrics/test_system_metrics_logging.py [hunk 2]
```diff
@@ -59,20 +60,23 @@ def test_manual_system_metrics_monitor():
...
+def test_automatic_system_metrics_monitor(wait_time, exit_wait_time, polling_freq):
```
Comment (harupy) [inline tests/system_metrics/test_system_metrics_logging.py:66]:
```suggestion
@pytest.mark.parametrize("x", range(1000))
def test_automatic_system_metrics_monitor(x, wait_time, exit_wait_time, polling_freq):
```

to run this test 1000 times to make sure it's not flaky.

**Comment by harupy** [comment]
@BenWilson2 can you attached the URL of the failed job in the PR description?

**Comment by harupy** [comment]
I filed https://github.com/mlflow/mlflow/pull/13692 last week. Can we use the same approach?

File: tests/system_metrics/test_system_metrics_logging.py (hunk 2 above)
Comment (harupy) [reply tests/system_metrics/test_system_metrics_logging.py:66]: We can remove this once we confirm the test is not flaky.

**Comment by BenWilson2** [comment]
> I filed #13692 last week. Can we use the same approach?

I just saw that this morning. I'm going to modify this test to use an identical approach. Thanks for the link / pointer :)

**Comment by BenWilson2** [comment]
Flakiness validation run: https://github.com/mlflow/mlflow/actions/runs/11822459686/job/32939584620?pr=13780

**Comment by BenWilson2** [comment]
Unflake validation for the mechanism that this PR targets (not returning sufficient step information for system metrics in Windows): https://github.com/mlflow/mlflow/actions/runs/11826525985?pr=13773 

There is one additional flake occurrence that is related to releasing the thread for the SystemMetricsMonitor via the automated interface out of the 200 tests iterations that can be handled in a separate PR as the mechanism is unrelated to the target of the fix in this PR.

File: tests/system_metrics/test_system_metrics_logging.py [hunk 3]
```diff
@@ -17,7 +17,22 @@ def disable_system_metrics_logging():
     mlflow.set_system_metrics_node_id(None)
 
 
-def test_manual_system_metrics_monitor():
+@pytest.fixture
```
[the remark this answers is not in the document]
Comment (BenWilson2) [reply tests/system_metrics/test_system_metrics_logging.py]: Not at all :) Changed to just a standard function to reduce the complexity of that context wrap

File: tests/system_metrics/test_system_metrics_logging.py [hunk 4]
```diff
@@ -17,7 +17,16 @@ def disable_system_metrics_logging():
     mlflow.set_system_metrics_node_id(None)
 
 
+def wait_for_condition(condition_func, timeout=10, check_interval=1):
+    for _ in range(timeout // check_interval):
+        if condition_func():
+            return True
+        time.sleep(check_interval)
+    return False
```
Comment (harupy) [inline tests/system_metrics/test_system_metrics_logging.py:25]: /ai add type hints.

File: tests/system_metrics/test_system_metrics_logging.py [hunk 5]
```diff
@@ -17,7 +17,16 @@ def disable_system_metrics_logging():
     mlflow.set_system_metrics_node_id(None)
 
 
+def wait_for_condition(condition_func, timeout=10, check_interval=1):
+    for _ in range(timeout // check_interval):
+        if condition_func():
+            return True
```
Comment (harupy) [inline tests/system_metrics/test_system_metrics_logging.py:23]: Do we need to return True/False?

Comment (BenWilson2) [reply tests/system_metrics/test_system_metrics_logging.py]: definitely. Updated to make this cleaner

File: tests/system_metrics/test_system_metrics_logging.py [hunk 6]
```diff
@@ -17,7 +18,19 @@ def disable_system_metrics_logging():
     mlflow.set_system_metrics_node_id(None)
 
 
+def wait_for_condition(
+    condition_func: Callable[..., Any], timeout: int = 10, check_interval: int = 1
```
Comment (harupy) [inline tests/system_metrics/test_system_metrics_logging.py:22]:
```suggestion
    condition_func: Callable[[], Any], timeout: int = 10, check_interval: int = 1
```

nit

**Review by harupy — APPROVED** [review]
LGTM once https://github.com/mlflow/mlflow/pull/13773/files#r1841516680 is resolved!

**Pull request merged**